File: src/analysis/ChemicalConversion.py
#  Copyright (C) 2016
#      Jakub Krajniak (jkrajniak at gmail.com)
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
******************************************
**espressopp.analysis.ChemicalConversion**
******************************************

Computes the ration of particles of given type with respect to the total number of particles.

.. math::

   x = n_t / n_{total}


where :math:`n_t` is a number of particles of type :math:`t` and :math:`n_{total}` is an input
parameter that in principle is a total number of particles of given type.

The usage is within the chemical reaction framework when during the reaction, the particles
change the types.

Two kinds of observables are implemented, basic one that counts the number of particles of given
type and second based on the sequence of types.

.. function:: espressopp.analysis.ChemicalConversion(system, particle_type, total_count)

            :param system: The system object
            :type system: espressopp.System
            :param particle_type: The particle type.
            :type particle_type: int
            :param total_count: The total number of particles of given type.
            :type total_count: int

.. function:: espressopp.analysis.ChemicalConversionTypeSequence(system, particle_group, type_seq, total_count)

            :param system: The system object
            :type system: espressopp.System
            :param particle_grouo: The particle type.
            :type particle_type: espressopp.ParticleGroup
            :param type_seq: The sequence of types to look for.
            :type type_seq: list
            :param total_count: The total number of molecules of given type sequence.
            :type total_count: int


Usage of ChemicalConversionTypeSequence
========================================

>>>> p = espressopp.ParticleGroup(system.storage)
>>>> [p.add(x) for x in range(1, 100)]   # creates a group of particles with ids 1...99
>>>> observable = espressopp.analysis.ChemicalConversionTypeSequence(system, p, [1, 2, 3], 100)

"""

from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.analysis.Observable import *  # NOQA
from _espressopp import analysis_ChemicalConversion
from _espressopp import analysis_ChemicalConversionTypeSequence


class ChemicalConversionLocal(ObservableLocal, analysis_ChemicalConversion):
    """The (local) compute of conversion."""
    def __init__(self, system, particle_type, total_count):
        if pmi.workerIsActive():
            cxxinit(self, analysis_ChemicalConversion, system, particle_type, total_count)

class ChemicalConversionTypeSequenceLocal(ObservableLocal, analysis_ChemicalConversionTypeSequence):
    """The (local) compute of conversion."""
    def __init__(self, system, particle_group, type_sequence, total_count):
        if pmi.workerIsActive():
            cxxinit(self, analysis_ChemicalConversion, system, particle_group, total_count)

            self.cxxclass.set_sequence(type_sequence)

if pmi.isController:
    class ChemicalConversion(Observable):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls='espressopp.analysis.ChemicalConversionLocal',
        )

    class ChemicalConversionTypeSequence(Observable):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls='espressopp.analysis.ChemicalConversionTypeSequenceLocal',
            pmicall=['set_sequence']
        )

File: src/analysis/ChemicalConversion.cpp
/*
  Copyright (C) 2015
    Jakub Krajniak (jkrajniak at gmail.com)

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include "ChemicalConversion.hpp"

using namespace espressopp;  //NOLINT

namespace espressopp {
namespace analysis {

real ChemicalConversion::compute_real() const {
  System& system = getSystemRef();
  CellList realCells = system.storage->getRealCells();

  longint local_count = 0;
  for (iterator::CellListIterator cit(realCells); !cit.isDone(); ++cit) {
    if (cit->type() == p_type)
      local_count++;
  }

  longint global_count = 0;
  boost::mpi::all_reduce(*getSystem()->comm, local_count, global_count, std::plus<longint>());

  real value = global_count;
  if (!absolute_value_)
    value = global_count / total_value;
  // Send value via signal.
  onValue(value);

  return value;
}

void ChemicalConversion::registerPython() {
  using namespace espressopp::python;  //NOLINT
  class_<ChemicalConversion, bases<Observable>, boost::noncopyable>
    ("analysis_ChemicalConversion",
        init< shared_ptr<System>, longint, longint >())
        .def(init<shared_ptr<System>, longint>());
}

real ChemicalConversionTypeSequence::compute_real() const {
  System& system = getSystemRef();
  CellList realCells = system.storage->getRealCells();

  longint local_count = 0;
  std::map<longint, longint> local_pid_type;
  for (ParticleGroup::iterator it = particle_group_->begin(); it != particle_group_->end(); ++it) {
    local_pid_type.insert(std::make_pair(it->id(), it->type()));
  }

  // Do the computation on root node as we have to match type sequence and
  // some particles can be splited among nodes.
  std::vector<std::map<longint, longint> > global_pid_type;

  real result = 0.0;
  if (system.comm->rank() == 0) {
    mpi::gather(*(system.comm), local_pid_type, global_pid_type, 0);

    std::map<longint, longint> global_seq;
    for (std::vector<std::map<longint, longint> >::iterator it = global_pid_type.begin();
         it != global_pid_type.end(); ++it ) {
      global_seq.insert(it->begin(), it->end());
    }

    // Flying window, looking for the sequence.
    size_t window_size = type_seq_.size();
    size_t particle_number = global_seq.size();
    if (particle_number % window_size != 0)
      throw std::runtime_error("Number of particles does not corresponds to the sequence length");

    longint counter = 0;

    for (std::map<longint, longint>::iterator it = global_seq.begin(); it != global_seq.end();) {
      bool valid = true;
      for (size_t j = 0; j < window_size; j++, ++it) {
        if (it->second != type_seq_[j] && valid)
          valid = false;
      }
      if (valid)
        counter++;
    }

  } else {
    mpi::gather(*(system.comm), local_pid_type, global_pid_type, 0);
  }

  mpi::broadcast(*(system.comm), result, 0);

  // Send value via signal.
  onValue(result);

  return result;
}

void ChemicalConversionTypeSequence::registerPython() {
  using namespace espressopp::python;  //NOLINT
  class_<ChemicalConversionTypeSequence, bases<Observable>, boost::noncopyable>
      ("analysis_ChemicalConversionTypeSequence",
       init< shared_ptr<System>, shared_ptr<ParticleGroup>, longint >())
      .def("set_sequence", &ChemicalConversionTypeSequence::setSequence);
}


}  // end namespace analysis
}  // end namespace espressopp
